Concatenate string literals in get_define

behlec opened the pull request:
If get_define returns a list of string_literals concatenate them into
one string.

Fixes #3663

nirbheek commented:
I realize that I said in the original issue that we could use `ast.literal_eval`, but I now think it's a bad idea because it will evaluate a lot more things than we expect, and might cause bugs because of that. Python syntax != C syntax.

Also, please add a test for it while you're at it :)

nirbheek commented:
In fact, looking at the test results, this already breaks other things: https://travis-ci.org/mesonbuild/meson/jobs/390168775

behlec commented:
Well, that would have been to easy :) I will work on it.

behlec commented:
There is a first possible implementation. There is still some work to do but I think one can get the idea.

CCompiler has now a method `concatenate_string_literals` which will search for string literals separated by spaces and join them.
`Compiler.get_define` has now an new keyword argument `concatenate_string_literals`. If it is set to `True`, it will run the result through `concatenate_string_literals` and return this result.
So far `concatenate_string_literals` is not publicly available.

Is that an acceptable approach?
Some TODOs:
* I will have to write more tests.
* Test if concatenate_string_literals is supported for this compiler, maybe #3731 fixes this.

jpakkane commented:
Could this be made to work automatically so there is no need for the new kwarg?

nirbheek commented:
> Could this be made to work automatically so there is no need for the new kwarg?

Yes, I think that string concatenation is an implementation detail of C defines that the user should not need to care about. What the user should get is what they expect to get when they use that define in C.

behlec commented:
I will write tests, to ensure there should be no problems in case the define is something strange.
If nothing shows up I will make it automatically.

behlec commented:
I have added tests and removed the kwarg. I added a sentence on the behavior in the reference manual.

Should I add an entry in the release notes as well?

nirbheek commented on test cases/common/140 get define/concat.h:
```diff
@@ -0,0 +1,23 @@
+#define TEST_CONCAT_1 \
+    "ab"              \
+    "cd"              \
+    "ef"
```
These are some nice tests. Could you also add a `""` at the end of this to test that it gets removed?

nirbheek commented on docs/markdown/Reference-manual.md:
```diff
@@ -1571,6 +1571,8 @@ the following methods:
 - `get_define(definename)` returns the given preprocessor symbol's
   value as a string or empty string if it is not defined.
+  The method will concatenate string literals as the compiler would.
+  E.g. `"a" "b"` will become `"ab"`.
```
This should say "Starting with *0.47.0*, the method"

nirbheek commented on mesonbuild/compilers/c.py:
```diff
@@ -933,6 +934,16 @@ def has_multi_link_arguments(self, args, env):
 
+    @staticmethod
+    def concatenate_string_literals(s):
+        pattern = re.compile(r'(?P<pre>.*([^\\]")|^")(?P<str1>([^\\"]|\\.)*)"\s+"(?P<str2>([^\\"]|\\.)*)(?P<post>".*)')
```
A regex is hard to read, but I feel like in this case it's the best solution.

nirbheek commented on docs/markdown/Release-notes-for-0.47.0.md:
```diff
@@ -21,3 +21,9 @@ Added the function `subdir_done()`. Its invocation exits the current script at
+
+After obtaining the value of a preprocessor symbol consecutive string literals
+are merged into a single string literal.
+For example a preprocessor symbol's value `"ab" "cd"` is returned as `"abcd"`.
```
Normally we ask people to put this in the `snippets/` subdir to avoid conflicts, but seeing as how we missed this once already, I guess it's ok :p

nirbheek commented:
This is in that grey area between feature and bugfix, so I'm going to err on the side of caution and pretend it's a new feature and wait for @jpakkane's approval. :)

jpakkane commented:
:shipit:
